File: yandex_winter_intern_day_2024/5. Округление/solution_.py
import sys

from typing import List

def main():
    n, x = map(int, input().split())
    a: List[int] = list(map(int, input().split()))

    total_a: int = sum(a)
    b: List[float] = [(x * ai) / total_a for ai in a]
    c: List[int] = [int(bi) for bi in b]
    
    current_sum: int = sum(c)
    diff: int = x - current_sum
    
    if diff == 0:
        print(*c)
        return
    
    indices: List[int] = sorted(range(n), key=lambda i: b[i] - c[i], reverse=diff > 0)

    for i in range(abs(diff)):
        if diff > 0:
            c[indices[i]] += 1
        else:
            c[indices[i]] -= 1
    
    print(*c)


if __name__ == '__main__':
    main()

File: yandex_winter_intern_day_2024/5. Округление/solution.cs
using System;
using System.Linq;


// 1. If the current sum is less than x, then we need to increase some c_i. 
// We choose the numbers with the largest fractional part and increase them by 1.
// 2. If the current sum is greater than x, then we need to decrease some c_i. 
// We choose the numbers with the smallest fractional part and decrease them by 1.

class Program 
{
    public static void Main(string[] args)
    {
        string[] line1 = Console.ReadLine().Split();
        int n = int.Parse(line1[0]);
        long x = int.Parse(line1[1]);

        long[] a = Console.ReadLine().Split().Select(long.Parse).ToArray();

        long totalA = a.Sum(e => (long)e);

        double[] b = a.Select(ai => (double)(x * ai) / totalA).ToArray();

        long[] c = b.Select(bi => (long)bi).ToArray();

        long currentSum = c.Sum();
        long diff = x - currentSum;

        if (diff == 0)
        {
            Console.WriteLine(string.Join(" ", c));
            return;
        }

        int[] indices;

        if (diff > 0)
            indices = Enumerable.Range(0, n).OrderByDescending(i => b[i] - c[i]).ToArray();
        else
            indices = Enumerable.Range(0, n).OrderBy(i => b[i] - c[i]).ToArray();

        for (long i = 0; i < Math.Abs(diff); i++)
        {
            if (diff > 0)
                c[indices[i]]++;
            else
                c[indices[i]]--;
        }

        Console.WriteLine(string.Join(" ", c));
    }
}